packages/babel-plugin-proposal-decorators/CONTRIB.md

```diff
@@ -0,0 +1,340 @@
+These are notes about the implementation of the 2020-09 decorators transform.
+The implementation's goals are (in descending order):
+
+1. Being accurate to the actual proposal (e.g. not defining additional
+  properties unless required, matching semantics exactly, etc.). This includes
+  being able to work properly with private fields and methods.
+2. Transpiling to a very minimal and minifiable output. This transform will
+  affect each and every decorated class, so ensuring that the output is not 10x
+  the size of the original is important.
+3. Having good runtime performance. Decoration output has the potential to
+  drastically impact startup performance, since it runs whenever a decorated
+  class is defined. In addition, every instance of a decorated class may be
+  impacted for certain types of decorators.
+
+All of these goals come somewhat at the expense of readability and can make the
+implementation difficult to understand, so these notes are meant to document the
+motivations behind the design.
+
+## Overview
+
+Given a simple decorated class like this one:
+
+```js
+@dec
+class Class {
+  @init:dec a = 123;
+
+  @dec static #b() {
+    console.log('foo');
+  }
+
+  @init:dec
+  accessor #c = 456;
+}
+```
+
+It's output would be something like the following:
+
+```js
+import { applyDecs } from '@babel/helpers';
+
+let initA, finishA, callB, initC, getC, setC, finishC;
+
+let Class;
+class _Class {
+  static {
```

Comment:
We can't group decorators into a single static block when DecoratorMemberExpression relies on execution orders e.g.

```js
class C {
  @(desc()) static q; // undefined is not a function
  static { desc = () => randomDecorator }
  @(desc()) static p;
}
```

Alternatively we can prepend static block to each element, preserving the execution order.

Reply:
The semantics here are a bit ambiguous. Consider:

```js
class C {
  static q = this.m();
  static { desc = () => randomDecorator }
  @(desc()) static m() {}
}
```

`m()` will already be defined on `C` when `q` is defined, so this will work. The current spec suggests that `m()` should be the _decorated_ version of `m`, and this is indeed a better DX, because then there is no point where a user can accidentally observer/use the _undecorated_ version of `m()`. Decorators are also evaluated all at once currently, when the class itself is defined and _before_ class fields are defined. So I believe that decorators should be evaluated _before_ class fields/static blocks run.

Reply:
```js
class C {
  static q = this.m();
  static { desc = () => randomDecorator }
  @(desc()) static m() {}
}
```

According to [ClassDefinitionEvaluation](https://tc39.es/ecma262/#sec-runtime-semantics-classdefinitionevaluation),
the initializer `this.m()` was wrapped into a function closure by [ClassFieldDefinitionEvaluation](https://tc39.es/ecma262/#sec-runtime-semantics-classfielddefinitionevaluation)(step 3.e) in step 25, and then evaluated by [DefineField](https://tc39.es/ecma262/#sec-definefield)(step 3.a) in step 31.a.i, at which all the methods have been defined, static blocks have been executed.

Now let's turn to the modified ClassDefinitionEvaluation in https://arai-a.github.io/ecma262-compare/?pr=2417

```js
class C {
  @(desc()) static q; // undefined is not a function
  static { desc = () => randomDecorator }
  @(desc()) static p;
}
```

Here, the first `desc()` is evaluated by DecoratorEvaluation (step 2) via DecoratorListEvaluation and ClassElementEvaluation in step 27.b of ClassDefinitionEvaluation (was 25.b in base spec), at which only the class methods before current element are defined, class fields before current element are initialized, static blocks before current element are evaluated.

> Decorators are also evaluated all at once currently, when the class itself is defined and before class fields are defined. So I believe that decorator expressions should be evaluated before class fields/static blocks run.

IIRC the current spec does not reflect that. We can add a step before step 27 to run DecoratorListEvaluation for each class element, and then initialize all the class elements. So in the example above, DecoratorListEvaluation on `desc()` always throws undefined is not a function regardless of the placement.

Reply:
> IIRC the current spec does not reflect that. We can add a step before step 27 to run DecoratorListEvaluation for each class element, and then initialize all the class elements. So in the example above, DecoratorListEvaluation on desc() always throws undefined is not a function regardless of the placement.

Moreso I meant that there is no way for _user_ code to observe any time between when two decorators are applied, similar to how it is not possible for user code to observe any time between when two methods are assigned to the prototype of a class. 

The intent was that all decorator evaluations occur as class elements are evaluated, which includes static fields. Then, in a subsequent step, static fields assigned and static blocks are executed. I _believe_ the spec reflects this, but will need to be sure when I rebase to include the changes for static blocks. 

All that said, I believe it is still correct to polyfill by running decorators in a static block at the top of the class. While not strictly the same timing as in the spec, from a user standpoint it will be identical, since decorators do not have access to the class itself and cannot observe a partially defined class.

Reply:
Oh I see. I gave the wrong example, I mentally thought the static block are evaluated immediately.

On the execution orders, I come up with a new example from computed keys:

```js
class C {
  @(f()) static q;
  [g()];
  @(h()) static p;
}
```

The evaluation order should be f, g, h.

Now if we group decorated fields into a single static block:
```js
class C {
  static {
    applyDec(
      this, [], [
       [[f(), 0], 0, 'q'],
       [[h(), 0], 1, 'p']
      ],
      []
    );
  }
  static q;
  [g()];
  static p;
}
```
The evaluation order now becomes g, f, h. To mitigate this, we can memoize the computed keys and decorate expression according to the order, the single static block approach should still work.

Reply:
Ah yes, I agree there that is correct, and sounds good, I can add that to the implementation